### DISCUSSION TITLE
How to implement Single Table Inheritance (STI) when using decorators

### DISCUSSION THREAD

**Discussion by n-xt** [CodeDredd/pinia-orm#519] [Q&A] [answered]
The manual shows how to use Single Tabel Inheritance using the static fields() method and using the spread operator calling super.fields(). But what when I am using decorators?

**Comment by CodeDredd** [comment]
If you just want to have a simple extend then look there: https://github.com/CodeDredd/pinia-orm/discussions/325

If you use types then yes i should have included an decorator example in the docs.
Here is an example:
````ts
    class Animal extends Model {
      static entity = 'animals'

      @Attr(null) declare id: number | null
      @Attr('animal') declare type: string

      static types() {
        return {
          animal: Animal,
          dog: Dog,
        }
      }
    }

    class Dog extends Animal {
      static entity = 'dogs'

      static baseEntity = 'animals'

      static fields() {
        return {
          ...super.schemas[super.entity],
        }
      }

      @Attr('dog') declare type
      @Attr('terrier') declare race: string
    }
````

**Comment by n-xt** [comment, marked as answer]
Thanks, I am still facing issues. I have separated the models into their own files. You mentioned that there could be circular reference error - but the link in the manual does not work. How should this be handled? Maybe there are some more issues?

Sandbox: https://codesandbox.io/s/single-table-inheritance-demo-21rukn?file=/src/App.vue

**Reply by CodeDredd** [reply]
Yeah there seem to be some issues. I got it to work, but its not yet working the way you wrote it.
I had to change the save repo with passing the type because it doesnt work yet to use a child repo and having the right type.
Look here: https://codesandbox.io/s/single-table-inheritance-demo-forked-by5l37?file=/src/App.vue

**Reply by n-xt** [reply]
Okay, I got it working now as I wanted:

- Querying the parent model "Resource" returns ALL data sets
- Querying a sub-model (Document / Image) only returns the data sets of that type
- I implemented each sub-model in its own file

Look here: https://codesandbox.io/s/single-table-inheritance-demo-21rukn?file=/src/App.vue

Thanks!

**Reply by CodeDredd** [reply]
Great to help you out although its not that way it should be....i am still wondering how it works. you just commented the `types` out....Have to rethink all this.

....
